draft-irtf-cfrg-hpke.md

```diff
@@ -1335,6 +1356,7 @@ the given mode satisfies the property.
 | Auth    | y            | y           | y            |
 | AuthPSK | y            | y           | y            |
 
+This analysis assumes that the KEM has no implicit errors.
```

Comment:
(I may just be misreading this. I'm not entirely clear on which errors are considered explicit vs implicit here.) It sounds like this sentence, with the rest of the paragraph, is saying that the analysis only holds for DH-based KEMs because non-DH-based KEMs may have implicit errors. But the text above suggests that DH-based KEMs *do* have implicit errors.

Reply:
It's... tricky :-) I'll leave it to @blipp to clarify precisely what is true about the analysis.

Reply:
This preliminary analysis is only about HPKE with DHKEM. DHKEM is “perfectly correct”, i.e. a valid encapsulation will be successfully decapsulated with probability 1. So, in some sense the analysis indeed assumes that there is no implicit error, but also it is not really an assumption, because we know that DH works that way.

I'd suggest that we do not add this sentence here, but instead add a sentence two paragraphs further down, in the paragraph that discusses the composition theorems proved in the Eurocrypt paper:

> Further, [ABHKLR20] proves composition theorems, showing that HPKE's Auth mode fulfills the security notions of authenticated public key encryption for all KDFs and AEAD schemes specified in this document, given any authenticated KEM satisfying the previously defined security notions for authenticated KEMs. **The theorems assume that the KEM is perfectly correct; it could easily be adapted to work with KEMs that have a non-zero negligible probability for decryption failure.** The assumptions on the KDF are that Extract() and Expand() can be modeled as pseudorandom functions wherein the first argument is the key, respectively. The assumption for the AEAD is IND-CPA and IND-CTXT security.

I will see if I can actually already adapt the composition proofs in CryptoVerif that way and will report back, but I think this is just a technical detail. It boils down to adding one game hop that turns the non-perfectly-correct KEM into a perfectly correct KEM, and accounting for the (negligible) probability that the adversary can detect this change.

On a related note: both analyses do _not_ explicitly model the all-zero/point-at-infinity check of the DH result. However, the theorems are valid for _both_ implementations that do or do not contain this check. (the reason is that it is predictable for the adversary if the DH result will be all-zero, and therefore the adversary could do this check itself before querying the KEM oracle)